app/src/main/java/com/example/criminalintentapp/presentation/fragments/crime_detail/CrimeFragment.kt

```diff
@@ -116,6 +142,59 @@ class CrimeFragment : Fragment(R.layout.fragment_crime), FragmentResultListener
                 activity?.supportFragmentManager?.popBackStack()
             }
         }
+
+        reportButton.setOnClickListener {
+            Intent(Intent.ACTION_SEND).apply {
+                type = "text/plain"
+                putExtra(Intent.EXTRA_TEXT, getCrimeReport())
+                putExtra(
+                    Intent.EXTRA_SUBJECT,
+                    getString(R.string.crime_report_subject)
+                )
+            }.also { intent ->
+                val chooserIntent = Intent.createChooser(intent, getString(R.string.send_report))
+                startActivity(chooserIntent)
+            }
+        }
+
+        suspectButton.apply {
+            val pickContactIntent =
+                Intent(Intent.ACTION_PICK, ContactsContract.Contacts.CONTENT_URI)
+
+            setOnClickListener {
+                resultLauncher.launch(pickContactIntent)
+            }
+        }
+    }
+
+    private fun onActivityResult(result: ActivityResult) {
+        if (result.resultCode != Activity.RESULT_OK) {
+            Log.d(TAG, "There is no result - onActiviyResult")
+            return
+        } else {
+            val contactUri: Uri? = result.data?.data
+            //Specify which fields you want your query to return values for
+            val queryFields = arrayOf(ContactsContract.Contacts.DISPLAY_NAME)
+            //Perform your query - the contactUri is like a "where" clause here
+            val cursor = contactUri?.let {
+                requireActivity().contentResolver
+                    .query(it, queryFields, null, null, null)
+            }
+            cursor?.use {
+                //Verify cursor contains at least one result
```

Comment:
The code inside the brackets seems to be the kind of logic that could also be placed inside a private function. Would help a lot with code cleanliness